### PR TITLE
Keccak circuit implementation

### PR DESCRIPTION
**Pull request by CPerezz**
This PR implements the entire keccak permutation circuit (No padding) to be used as a gadget or a final circuit.

Resolves: #218
Resolves: #273

### CODE REVIEW

File: keccak256/src/circuit.rs
Comment (ChihChengLiang) [inline keccak256/src/circuit.rs:346]:
```suggestion
                self.keccak_conf.rho_config.load(layouter)?;
                self.table.load(layouter)?;
```

Comment (CPerezz) [reply keccak256/src/circuit.rs:346]: I'm stupid...

**Comment by ChihChengLiang** [comment]
@CPerezz Can we do a rebase before the review?

File: keccak256/src/gates/absorb.rs
Comment (therealyingtong) [inline keccak256/src/gates/absorb.rs:57]:
```suggestion
                // This is boolean-constrained outside of `AbsorbConfig` by `MixingConfig`.
                let flag = meta.query_advice(state[NEXT_INPUTS_LANES], Rotation::cur());
```

File: keccak256/src/gates/absorb.rs
Comment (therealyingtong) [inline keccak256/src/gates/absorb.rs:83]:
```suggestion
    fn assign_next_inp_and_flag(
```

File: keccak256/src/gates/absorb.rs
Comment (therealyingtong) [inline keccak256/src/gates/absorb.rs:95]:
```suggestion
            // Assign only first 17 values.
            if x >= 3 && y >= 1 {
                break;
            }
```

File: keccak256/src/gates/absorb.rs
Comment (therealyingtong) [inline keccak256/src/gates/absorb.rs:110]:
```suggestion
        // Assign flag at last column(17th).
```

File: keccak256/src/gates/absorb.rs
Comment (therealyingtong) [inline keccak256/src/gates/absorb.rs:100]: The offset increment is now done outside this helper.

```suggestion
        // Assign next_mixing.
```

File: keccak256/src/gates/base_conversion.rs
Comment (therealyingtong) [inline keccak256/src/gates/base_conversion.rs:151]: Is this meant to be commented out?

File: keccak256/src/gates/mixing.rs
Comment (therealyingtong) [inline keccak256/src/gates/mixing.rs:166]:
```suggestion
                    || "witness negated is_mixing",
```

File: keccak256/src/gates/mixing.rs
Comment (therealyingtong) [inline keccak256/src/gates/mixing.rs:172]:
```suggestion
                let negated_flag = (cell, F::from(flag_bool as u64));

                Ok((flag, negated_flag))
```

File: keccak256/src/gates/mixing.rs
Comment (therealyingtong) [inline keccak256/src/gates/mixing.rs:437]:
```suggestion
                    || "Mixing witness assignment",
```

**Review by therealyingtong — CHANGES_REQUESTED** [review]
First-pass review.

File: keccak256/src/gates/mixing.rs
Comment (ChihChengLiang) [inline keccak256/src/gates/mixing.rs:73]: We should split 3 polys here. Otherwise, the prover can witness values not satisfying individual polys but satisfying the summation poly.
Proof: let `x=flag`, `y=negated_flag`
flag_consistency: `x + y - 1`
bool_constraint for x: `x(1-x)`
bool_constraint for y: `y(1-y)`
summation: `x + y - 1 + x(1-x) + y(1-y) = 0`
falsifying example: `x=2 and y=1`

```suggestion
            [q_flag.clone() * flag_consistency, q_flag.clone() * bool_constraint(flag), q_flag * bool_constraint(negated_flag))]
```

**Review by ChihChengLiang — COMMENTED** [review]
First pass on gates

File: circuit-benchmarks/src/keccak_permutation.rs
Comment (CPerezz) [reply circuit-benchmarks/src/keccak_permutation.rs:87]: No, you're right! Thanks for the catch!

File: circuit-benchmarks/src/keccak_permutation.rs
Comment (CPerezz) [reply circuit-benchmarks/src/keccak_permutation.rs:263]: No, indeed that's just a consistency check. I was even planning to remove it as it does not provide any helpful results that we don't have previously.

File: keccak256/src/arith_helpers.rs
Comment (CPerezz) [reply keccak256/src/arith_helpers.rs:246]: Will file an issue for the `AssignedCell` to work on it as soon as this PR is merged.

Comment (CPerezz) [reply keccak256/src/arith_helpers.rs:246]: Filled #323

File: keccak256/src/circuit.rs
Comment (CPerezz) [reply keccak256/src/circuit.rs:32]: Yes. 

`base_conv_activator` is always `1` since it's needed in order to activate the base conversion gadget in the loop of the keccak permutation. (Where Mixing Step is not called).

Instead, `is_mixing_flag` is just used on that last iteration when we perform the `Mixing`.

File: keccak256/src/circuit.rs
Comment (CPerezz) [reply keccak256/src/circuit.rs:196]: No, it's not as I was planning to convert this to a Fixed column so that it does not need to be witnessed nor constrained.

WDYT? If makes sense I can make the changes now.

File: keccak256/src/gates/base_conversion.rs
Comment (CPerezz) [reply keccak256/src/gates/base_conversion.rs:151]: It is. We reviewed that with @ChihChengLiang and this leads to a Permutation err as we copy from different cells different rounds.

We should maybe look into this a bit more.

File: keccak256/src/gates/mixing.rs
Comment (CPerezz) [reply keccak256/src/gates/mixing.rs:166]: Good catch!

File: keccak256/src/circuit.rs
Comment (CPerezz) [reply keccak256/src/circuit.rs:384]: Nope. You're correct!

File: keccak256/src/gates/mixing.rs
Comment (CPerezz) [reply keccak256/src/gates/mixing.rs:73]: That makes sense!!! Thanks for catching this!!!!

**Comment by Brechtpd** [comment]
https://github.com/appliedzkp/zkevm-circuits/blob/4732717f959cb407a8b8bf6514dbc79c55b64291/keccak256/src/gates/rho_checks.rs#L459-L479

The `step2_acc` constraint (with final degree 14) and especially the `step3_acc` constraint (with final degree 171!!!) make the extended degree of the keccak circuit to be 256x the size of the normal domain, which has an enormous impact on performance and memory use. It will definitely be worth it to lower the degree of these constraints by either doing the checks differently in a way that has a lower degree, or use some extra cells so that the full expression can be split up into multiple sub expressions of a smaller degree.

Removing just these two constraints makes the circuit have a much more reasonable extended domain of just x8 bigger. The max degree of 7 is then caused by lookups (which have expressions up to degree 3 as input, but because they are used in the lookup an extra degree of 4 is added with current halo2 with zk), so probably still possible to get it down further with some additional tweaks, but of course will have much less impact than fixing the two constraints above.

**Comment by ChihChengLiang** [comment]
@Brechtpd Nice catch for the performance bottleneck. Do you think using a lookup table check for step2_acc and step3_acc would improve the performance?

File: keccak256/src/gates/base_conversion.rs
Comment (ChihChengLiang) [reply keccak256/src/gates/base_conversion.rs:151]: per offline discussion with @CPerezz, the copy constraint is required, but it might be difficult to enable in this PR. We can enable it in a future PR.

File: keccak256/src/circuit.rs
Comment (therealyingtong) [reply keccak256/src/circuit.rs:32]: I see, just saw the `TODO` to change `_base_conv_activator` to a `Selector`.

**Review by therealyingtong — APPROVED** [review]
LGTM modulo some small suggestions.

Another nitpick: I think we should avoid prefixing variable names with an underscore `_`, because that makes the compiler not warn us when these variables are unused.

File: keccak256/src/circuit.rs
Comment (CPerezz) [reply keccak256/src/circuit.rs:54]: You are right it's not. 
Good catch!

File: keccak256/src/keccak_arith.rs
Comment (ChihChengLiang) [inline keccak256/src/keccak_arith.rs:132]: Do we accidentally add these?

```suggestion
        let padding_total = self.sponge.rate - (input.len() % self.sponge.rate);
```

File: keccak256/src/keccak_arith.rs
Comment (ChihChengLiang) [inline keccak256/src/keccak_arith.rs:146]:
```suggestion
        let padded_input: &[u8] = &[input, &padding].concat();
```

File: keccak256/src/keccak_arith.rs
Comment (ChihChengLiang) [inline keccak256/src/keccak_arith.rs:192]:
```suggestion
```

File: keccak256/src/keccak_arith.rs
Comment (ChihChengLiang) [inline keccak256/src/keccak_arith.rs:173]:
```suggestion
```

**Review by ChihChengLiang — COMMENTED** [review]
Found some undeleted comments

**Comment by CPerezz** [comment]
> Found some undeleted comments

Sorry @ChihChengLiang as mentioned in private, this was the result of pushing some debugging lines that I was not planning to push.

I've force-pushed so that the history is clean now.

**Pull request merged**
